File: classic/Code/ConvertSortedListtoBinarySearchTree.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode*sort(ListNode*head,ListNode*tail){
        if(head==tail)return NULL;
        int N=0;
        ListNode*tmp=head;
        while(tmp!=tail){
            N++;
            tmp=tmp->next;
        }
        int mid=N/2;
        tmp=head;
        for (int i=0;i<mid;i++){
            tmp=tmp->next;
        }
        TreeNode*tHead=new TreeNode(tmp->val);
        tHead->left=sort(head,tmp);
        tHead->right=sort(tmp->next,tail);
        return tHead;
        
    }
    TreeNode *sortedListToBST(ListNode *head) {
        return sort(head,NULL);
    }
};

File: classic/Code/ConvertSortedListtoBinarySearchTree.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; next = null; }
 * }
 */
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode convert(ListNode start, ListNode end){
        int n=0;
        ListNode tmp=start;
        if(start==end)return null;
        while(tmp!=end&&tmp!=null){
            tmp=tmp.next;
            n++;
        }
        int mid=n/2;
        tmp=start;
        for(int i=0;i<mid;i++){
            tmp=tmp.next;
        }
        TreeNode node=new TreeNode(tmp.val);
        node.left=convert(start,tmp);
        node.right=convert(tmp.next,end);
        return node;
    }
    public TreeNode sortedListToBST(ListNode head) {
        return convert(head,null);
    }
}
